app/models/upload.rb

```diff
@@ -316,6 +317,64 @@ def thumbnail_height
     get_dimension(:thumbnail_height)
   end
 
+  def dominant_color(calculate_if_missing: false)
+    val = read_attribute(:dominant_color)
+    if val.nil? && calculate_if_missing
+      calculate_dominant_color!
+      read_attribute(:dominant_color)
+    else
+      val
+    end
+  end
+
+  def calculate_dominant_color!(local_path = nil)
+    color = nil
+
+    if !FileHelper.is_supported_image?("image.#{extension}") || extension == "svg"
+      color = ""
+    end
+
+    if color.nil?
+      local_path ||=
+        if local?
+          Discourse.store.path_for(self)
+        else
+          Discourse.store.download(self).path
+        end
+
+      color = begin
+        data = Discourse::Utils.execute_command(
+          "nice",
+          "-n",
+          "10",
+          "convert",
+          local_path,
+          "-colors",
+          "1",
+          "-define",
+          "histogram:unique-colors=true",
+          "-format",
+          "%c",
+          "histogram:info:",
+          timeout: MAX_DOMINANT_COLOR_SECONDS
+        )
+
+        color = data[/#([0-9A-F]{6})/, 1]
+        raise "Unable to parse dominant color output:\n#{data}" if color.nil?
+        color
+      rescue => e
+        Discourse.warn_exception(e, message: "Error calculating dominant color for Upload id=#{id}")
+        ""
```

Comment:
Kind of related to https://github.com/discourse/discourse/pull/18248/files#r973821411, I wonder if we should differentiate  between the type of errors. Right now, we're give up on an image forever once it encounters any error during this process. It could be a typo in the code, or a timeout during the command.

Reply:
Tightened up the `rescue` so it only swallows ImageMagick errors (i.e. bad image input), and timeouts. Any other exceptions will raise as normal